src/memray/_memray/tracking_api.cpp

```diff
@@ -300,6 +310,97 @@ PythonStackTracker::popPythonFrame(PyFrameObject* frame)
     invalidateMostRecentFrameLineNumber();
 }
 
+void
+PythonStackTracker::installGreenletTraceFunctionIfNeeded()
+{
+    if (!s_greenlet_tracking_enabled || d_greenlet_hooks_installed) {
+        return;  // Nothing to do.
+    }
+
+    assert(PyGILState_Check());
+
+    RecursionGuard guard;
+
+    // Borrowed reference
+    PyObject* modules = PySys_GetObject("modules");
+    if (!modules) {
+        return;
+    }
+
+    // Borrowed reference
+    // Look directly at `sys.modules` since we only want to do something if
+    // `greenlet._greenlet` has already been imported.
+    PyObject* _greenlet = PyDict_GetItemString(modules, "greenlet._greenlet");
+    if (!_greenlet) {
+        return;
+    }
+
+    // Borrowed reference
+    PyObject* _memray = PyDict_GetItemString(modules, "memray._memray");
+    if (!_memray) {
+        return;
+    }
+
+    PyObject* ret = PyObject_CallMethod(
+            _greenlet,
+            "settrace",
+            "N",
+            PyObject_GetAttrString(_memray, "greenlet_trace_function"));
+    Py_XDECREF(ret);
+
+    if (!ret) {
+        // This might be hit from PyGILState_Ensure when a new thread state is
+        // created on a C thread, so we can't reasonably raise the exception.
+        PyErr_Print();
+        _exit(1);
+    }
+
+    // Note: guarded by the GIL
+    d_greenlet_hooks_installed = true;
+
+    static bool warned = false;
+    if (!warned) {
+        warned = true;
+
+        PyObject* ret = PyObject_CallMethod(_memray, "print_greenlet_warning", nullptr);
+        Py_XDECREF(ret);
+        if (!ret) {
+            PyErr_Print();
+            _exit(1);
+        }
+    }
+}
+
+void
+PythonStackTracker::handleGreenletSwitch(PyObject*, PyObject*)
+{
+    RecursionGuard guard;
+
+    // Clear any old TLS stack, emitting pops for frames that had been pushed.
+    if (d_stack) {
+        for (auto& frame : *d_stack) {
+            if (frame.state != FrameState::NOT_EMITTED) {
+                ++d_num_pending_pops;
+            }
+        }
+        d_stack->clear();
+        emitPendingPushesAndPops();
+    }
+
+    // Re-create our TLS stack from our Python frames, most recent last.
+    PyFrameObject* frame = PyEval_GetFrame();
```

Comment:
This is correct because if this is NULL for whatever reason we don't do anything, but it may be good to add an assert or early return to make it clearer.

Reply:
It's written to allow NULL, as you say. I'd rather add a comment than more code, I'll do that.